File: leetcode_ly/java/Tree/ValidateBinarySearchTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isValidBST(TreeNode root) {
        //看到二分查找树就要想到中序遍历
        List<Integer> list = new ArrayList<Integer>();
        inorder(root,list);
        for(int i = 1; i < list.size(); i++){
            if(list.get(i) <= list.get(i - 1)) return false;
        }
        return true;
    }
    
    public void inorder(TreeNode root,List<Integer> list){
        if(root == null) return;
        inorder(root.left,list);
        list.add(root.val);
        inorder(root.right,list);
        return;
    }
}

File: leetcode_ly/c++/Tree/ValidateBinarySearchTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> ret;

    bool isValidBST(TreeNode* root) {
        inorder(root);
        for(int i = 1; i < ret.size(); i++){
            if(ret[i] <= ret[i - 1]) return false;
            else continue;
        }
        return true;
    }
    
    void inorder(TreeNode* root){
        if(!root) return;
        inorder(root->left);
        ret.push_back(root->val);
        inorder(root->right);
    }
};